colibri/src/blockchain/node_inquirer.rs

```diff
@@ -0,0 +1,107 @@
+use std::sync::Arc;
+use std::collections::HashMap;
+use alloy::providers::DynProvider;
+use alloy::providers::ProviderBuilder;
+use log::{debug, error};
+use tokio::sync::RwLock;
+
+use crate::blockchain::{SupportedBlockchain, RpcNode};
+use crate::globaldb::GlobalDB;
+
+
+/// A struct that manages connections to EVM RPC nodes for a specific blockchain
+pub struct EvmNodeInquirer {
+    globaldb: Arc<GlobalDB>,
+    pub rpc_nodes: RwLock<Vec<RpcNode>>,
+    pub blockchain: SupportedBlockchain,
+    // Connected nodes
+    provider_mapping: RwLock<HashMap<RpcNode, Box<DynProvider>>>,
+}
+
+impl EvmNodeInquirer {
+    pub fn new(blockchain: SupportedBlockchain, globaldb: Arc<GlobalDB>) -> Self {
+        let instance = Self {
+            blockchain,
+            globaldb,
+            rpc_nodes: RwLock::new(Vec::new()),
+            provider_mapping: RwLock::new(HashMap::new()),
+        };
+
+        debug!("created EvmNodeInquirer for {}", blockchain.as_str());
+        instance
+    }
+
+    pub async fn update_rpc_nodes(&self) -> Result<(), String> {
+        let nodes = self.globaldb.get_rpc_nodes(self.blockchain)
+            .await
+            .map_err(|e| format!("Failed to get RPC nodes: {}", e))?;
+
+        *self.rpc_nodes.write().await = nodes;
+        Ok(())
+    }
+
+    pub async fn connect_to_node(&self, node: &RpcNode) -> Result<Arc<DynProvider>, String> {
```

Comment:
Add a docstring and rename this to explain what it does. A sugested name.
```suggestion
    pub async fn get_or_create_node_connection(&self, node: &RpcNode) -> Result<Arc<DynProvider>, String> {
```